openedx/features/course_experience/__init__.py

```diff
@@ -22,62 +23,61 @@ class DefaultTrueWaffleFlagNamespace(WaffleFlagNamespace):
-
-        Note: This is a similified version of the method it overrides, that
-        hard codes the default to True, and skips the call back used for
-        course overrides:
-        https://github.com/edx/edx-platform/blob/df9be8c678f8266e2e5710513c74deca14c4527c/openedx/core/djangoapps/waffle_utils/__init__.py#L229-L305
-
+        Override the default flag value in the absence of flag: if the
+        flag does not exist, it is assumed to be True (active). See the
+        parent class.
         """
         # Import is placed here to avoid model import at project startup.
         from waffle.models import Flag
 
-        # validate arguments
-        namespaced_flag_name = self._namespaced_name(flag_name)
-        value = self._cached_flags.get(namespaced_flag_name)
-        if value is None:
+        # determine if the flag is undefined in waffle
+        try:
+            Flag.objects.get(name=namespaced_flag_name)
+        except Flag.DoesNotExist:
+            # default to True if not defined
+            return True
 
-            # determine if the flag is undefined in waffle
-            try:
-                Flag.objects.get(name=namespaced_flag_name)
-            except Flag.DoesNotExist:
-                # default to True if not defined
-                value = True
+        return super()._get_flag_active_request(namespaced_flag_name, request)
 
-            if value is None:
-                request = crum.get_current_request()
-                if request:
-                    value = flag_is_active(request, namespaced_flag_name)
-                else:
-                    set_custom_attribute('warn_flag_no_request', True)
-                    # Return the default value if not in a request context.
-                    # Same as the original implementation
-                    self._set_waffle_flag_attribute(namespaced_flag_name, value)
-                    return True
 
-            self._cached_flags[namespaced_flag_name] = value
+class DefaultTrueCourseWaffleFlag(CourseWaffleFlag):
+    """
+    Override the default flag values by ignoring course-specific behaviour.
 
-        self._set_waffle_flag_attribute(namespaced_flag_name, value)
-        return value
+    WARNING: This is a temporary class that should not be used further. Instead, you should use the
+    edx_toggles.toggles.WaffleFlag class.
+    """
+    def __init__(self, waffle_namespace, flag_name, module_name=None):
+        if not isinstance(waffle_namespace, DefaultTrueWaffleFlagNamespace):
+            message = (
+                "DefaultTrueCourseWaffleFlag must only be associated to DefaultTrueWaffleFlagNamespace objects, not"
+                " {}"
+            ).format(waffle_namespace.__class__.__name__)
+            raise ValueError(message)
+
+        super().__init__(waffle_namespace, flag_name, module_name=module_name)
+
+    def _get_course_override_value(self, course_key):
+        return None
```

Comment:
Why are we disabling the ability to set course overrides on this?  I thought the only difference for this was to have a default of True instead of False if the waffle flag is not defined (which is in the namespace code).

Reply:
There is another difference: the DefaultTrueWaffleFlagNamespace skips the callback defined by the `check_before_waffle_callback` argument. In practice this is equivalent to skipping course-specific values.

Reply:
1. That was probably my bad while implementing this originally.
2. All of this was supposed to be very temporary, because these flags were in-progress of being removed, but I don't know when they will be removed.
3. As I stated earlier, I wonder if we should just reinstate the optional `default` argument, and allow this to be a simple CourseWaffleFlag with `default=True` as it was originally. The [ADR that had me drop this](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/waffle_utils/docs/decisions/0001-refactor-waffle-flag-default.rst) is no longer applicable really. The original toggle state report was gathering data directly from the db, and didn't have access to the default that was in code.  Now, instead, it relies on the endpoint data that has access to both the db and the default in the code.

Reply:
> That was probably my bad while implementing this originally.

I don't think so, as this behaviour is explicitly described by the class docstring: 

```
        Note: This is a similified version of the method it overrides, that
        hard codes the default to True, and skips the call back used for
        course overrides:
        https://github.com/edx/edx-platform/blob/df9be8c678f8266e2e5710513c74deca14c4527c/openedx/core/djangoapps/waffle_utils/__init__.py#L229-L305
```

> As I stated earlier, I wonder if we should just reinstate the optional default argument, and allow this to be a simple CourseWaffleFlag with default=True as it was originally.

In this PR I tried very hard to make the changes backward-compatible in order to avoid difficult decisions in the middle of the refactoring process. If you think this needs to be changed, maybe one of the stakeholders for this feature can step in and make a proposition?

Reply:
To document the decision from our discussion, this PR is blocked waiting on a new PR that updates:
* `COURSE_OUTLINE_PAGE_FLAG` => `DISABLE_COURSE_OUTLINE_PAGE_FLAG`
* `UNIFIED_COURSE_TAB_FLAG` => `DISABLE_UNIFIED_COURSE_TAB_FLAG`
Also, `DefaultTrueWaffleFlagNamespace` and `DefaultTrueCourseWaffleFlag` will be removed.

Reply:
Here is the PR: https://github.com/edx/edx-platform/pull/25440